usdigitalresponse/grants-ingest
# Additional max-split configurations for `SplitGrantsGovXMLDB`

## Description

This PR expands the ability to set limits on the number of records split from a Grants.gov XML database export, which was [originally introduced](https://github.com/usdigitalresponse/grants-ingest/commit/db256e68053736078b5fa8c9903d3f4b64927c3d#diff-8f3751bd003a96e924ff77bdcf8cf95aba7624a093d318d454d390e05b95ee32R127-R129) in https://github.com/usdigitalresponse/grants-ingest/pull/892.

With this PR's changes, max limits can also be configured for forecast- and/or opportunity-type grant records, which is useful for functional testing in local development environments where the goal is to ensure that a certain amount of each type of record gets processed. The original `MAX_SPLIT_RECORDS` environment variable is still supported, which imposes a hard limit on processing (that is, given a configuration like
```
MAX_SPLIT_OPPORTUNITY_RECORDS=10
MAX_SPLIT_FORECAST_RECORDS=10
MAX_SPLIT_RECORDS=5
```
processing will end once 5 records are split, even though 10 of each type of record have not been processed.
Therefore, this configuration:
```
MAX_SPLIT_OPPORTUNITY_RECORDS=5
MAX_SPLIT_RECORDS=10
```
is equivalent to this configuration:
```
MAX_SPLIT_OPPORTUNITY_RECORDS=5
MAX_SPLIT_FORECAST_RECORDS=5
MAX_SPLIT_RECORDS=10
```

In other words, assuming all three environment variables are configured, the process of splitting grant records will end as soon as `MAX_SPLIT_OPPORTUNITY_RECORDS` **and** `MAX_SPLIT_FORECAST_RECORDS` are reached, **or** `MAX_SPLIT_RECORDS`, whichever comes first.

Additionally, this PR exposes a new project-level Terraform input variable, `is_forecasted_grants_enabled`, which sets the variable of the same name for the `modules/SplitGrantsGovXMLDB` module. The default value is `false`, but it is set to `true` for local development environments in `terraform/local.tfvars`.

## Testing

1. Ensure that a fresh LocalStack environment is running with `docker compose`
2. Deploy to LocalStack by running `task local:from-scratch`
3. Kick off the pipeline by running `task local:invoke-DownloadGrantsGovDB`
4. After the `SplitGrantsGovXMLDB` pipeline step completes, check to ensure that the S3 bucket used to store prepared grants data contains 20 objects – 10 of which have a suffix of `/v2.OpportunitySynopsisDetail_1_0.xml` and another 10 with a suffix of `/v2.OpportunityForecastDetail_1_0.xml`.
5. Optional next steps:
  - Configure additional limits for testing as desired by setting the `max_split_grantsgov_opportunity_records`, `max_split_grantsgov_forecast_records`, and/or `max_split_grantsgov_records` Terraform input variables in `terraform/local.tfvars`. 
  - To deploy the new configuration to the existing LocalStack environment and test changes, run `task local:deploy`, and then proceed starting at Step 3.
  - To deploy the new configuration to a fresh LocalStack environment and test changes, run `docker compose restart` to reset LocalStack, and then proceed starting at Step 2.

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
